app/src/main/java/org/mozilla/tv/firefox/webrender/cursor/CursorModel.kt

```diff
@@ -0,0 +1,352 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.tv.firefox.webrender.cursor
+
+import android.annotation.SuppressLint
+import android.graphics.PointF
+import android.view.KeyEvent
+import android.view.MotionEvent
+import androidx.annotation.CheckResult
+import io.reactivex.Observable
+import io.reactivex.rxkotlin.Observables
+import io.reactivex.subjects.BehaviorSubject
+import io.reactivex.subjects.PublishSubject
+import org.mozilla.tv.firefox.ScreenControllerStateMachine
+import org.mozilla.tv.firefox.ScreenControllerStateMachine.ActiveScreen.WEB_RENDER
+import org.mozilla.tv.firefox.ext.isUriYouTubeTV
+import org.mozilla.tv.firefox.framework.FrameworkRepo
+import org.mozilla.tv.firefox.session.SessionRepo
+import org.mozilla.tv.firefox.utils.Direction
+import org.mozilla.tv.firefox.utils.DirectionHelper
+
+// Constants that can be tweaked to adjust cursor behavior
+private const val BASE_SPEED = 5f
+private const val MAX_VELOCITY = 25f
+private const val MS_TO_MAX_ACCELERATION = 200
+private const val MAX_SCROLL_VELOCITY = 13
+// Other constants
+private const val MAX_ACCELERATION = MAX_VELOCITY - BASE_SPEED
+private const val ACCELERATION_PER_MS = MAX_ACCELERATION / MS_TO_MAX_ACCELERATION
+private const val MS_PER_FRAME = 16
+private const val EDGE_OF_SCREEN_MARGIN = 1
+private const val NOT_RECENTLY_UPDATED = -1L
+
+data class HandledKeypress(val wasConsumed: Boolean, val simulatedTouch: MotionEvent?)
+
+sealed class CursorEvent {
+    /**
+     * This represents the user moving the cursor to the edge of the screen,
+     * attempting to scroll, and failing because the website has no more
+     * content in that direction.
+     */
+    data class ScrolledToEdge(val edge: Direction) : CursorEvent() // TODO how often does this happen?  Telemetry would be good
+    data class CursorMoved(val direction: Direction) : CursorEvent()
+}
+
+/**
+ * Calculates cursor movements.
+ *
+ * Remote key events are pushed here from the rest of the app. Actual cursor position changes are
+ * made when this class is called by [CursorView.onDraw]. This was done for performance reasons;
+ * see the comment on [mutatePosition] for more details.
+ */
+class CursorModel(
+    activeScreen: Observable<ScreenControllerStateMachine.ActiveScreen>,
+    frameworkRepo: FrameworkRepo,
+    sessionRepo: SessionRepo
+) {
+    // This is set early in the class lifecycle. Most methods short if it is not available
+    var screenBounds: PointF? = null
+    var webViewCouldScrollInDirectionProvider: (Direction) -> Boolean = { false }
+
+    private val directionKeysPressed = mutableSetOf<Direction>()
+    private var lastVelocity = 0f
+    private var lastUpdatedAtMS = NOT_RECENTLY_UPDATED
+    private var lastKnownCursorPos = PointF(0f, 0f)
+    private var cursorHasBeenCentered = false
```

Comment:
nit: `isInitialCursorPositionSet`? Without "initial position", it's unclear when the cursor gets centered and why.